src/overlay/Peer.cpp

```diff
@@ -992,6 +1017,17 @@ Peer::recvHello(Hello const& elo)
         return;
     }
 
+    mAddress = makeAddress(elo.listeningPort);
+    if (mAddress.isEmpty())
+    {
+        CLOG(WARNING, "Overlay") << "bad address in recvHello";
+        mDropInRecvHelloAddressMeter.Mark();
+        drop(ERR_CONF, "bad address", Peer::IGNORE_WRITE_QUEUE);
+        return;
+    }
+
+    updatePeerRecordAfterEcho();
```

Comment:
Its also for promotion. If some peer connected to us, we may want to promote it to outbound peer. So we connect to it, send our hello and receive hello from it. Then we need to store that this peere is outbound before we reject it because we are already connected to it. So it need to happen inside recvHello but before it drops in response to predicates isAuthenticated or isPending that are checked at the end of recvHello.

Reply:
still not following in the context of this commit: we don't drop a peer that early.

The check `if (getRole() == authenticatedIt->second->getRole())` ensures that we don't drop a peer already connected to us if all we're doing is promotion.

so we should always reach the call to `updatePeerRecordAfterAuthentication` in that situation.

Maybe I am missing something

Reply:
This condition was removed later: https://github.com/stellar/stellar-core/pull/1807/commits/2ab54458075ef21dd1dfd2e3c92c83431f76b796#diff-239e869bd0b671610b35830852c1cbadL1038

Ordering of changes will be fixed during rebase/squash phase.